selfdrive/locationd/magd.py

```diff
@@ -0,0 +1,221 @@
+    return calibration_params
+
+  def reset_angle_range(self, angle: Union[float, np.ndarray]) -> Union[float, np.ndarray]:
+    # reset angle range to [-pi, pi]
+    return (angle + np.pi) % (2 * np.pi) - np.pi
+
+  def get_calibrated_bearing(self, x: Union[float, np.ndarray], y: Union[float, np.ndarray], calibration_params: np.ndarray) -> Any:
+    x0, y0, l1, l2, theta, offset = calibration_params
+    x = x - x0
+    y = y - y0
+    x_new = x * np.cos(theta) + y * np.sin(theta)
+    y_new = -x * np.sin(theta) + y * np.cos(theta)
+    x_new, y_new = x_new / l1, y_new / l2
+    bearing = np.arctan2(y_new, x_new)
+    bearing = self.reset_angle_range(bearing - offset)
+    return bearing
+
+  def handle_log(self, which: str, msg: log.Event) -> None:
+    if which == "carState":
+      self.vego = msg.vEgo
+    elif which == "liveLocationKalman":
+      self.yaw = msg.orientationNED.value[2]
+      self.yaw_valid = msg.orientationNED.valid
+    elif which == "magnetometer":
+      if msg.source == log.SensorEventData.SensorSource.mmc5603nj:
+        raw_vals = np.array(msg.magneticUncalibrated.v)
+        self.raw_vals = (raw_vals[: 3] - raw_vals[3:]) / 2
+        if np.all(np.abs(self.raw_vals - self.past_raw_vals) < 2.0) and msg.magneticUncalibrated.status:
+          self.filtered_vals = np.array([f.update(v) for f, v in zip(self.filtered_magnetic, self.raw_vals, strict=True)])
+        if self.vego > MIN_VEL and self.yaw_valid:
+          self.point_buckets.add_point(self.filtered_vals[0], self.filtered_vals[2], self.yaw)
+        self.past_raw_vals = self.raw_vals
+
+  def get_msg(self, valid: bool = True, with_points: bool = False) -> log.Event:
+    msg = messaging.new_message('magnetometerCalibration')
+    msg.valid = valid
+    magnetometerCalibration = msg.magnetometerCalibration
+    magnetometerCalibration.version = VERSION
+
+    bearing, bearingValid = 0.0, False
+    if self.calibrated:
+      bearing = self.get_calibrated_bearing(self.filtered_vals[0], self.filtered_vals[2], self.calibrationParams)
+      bearingValid = True
+
+    magnetometerCalibration.calibrated = self.calibrated
+    magnetometerCalibration.calibrationParams = self.calibrationParams.tolist()
+    magnetometerCalibration.bearing = float(bearing)
+    magnetometerCalibration.bearingValid = bearingValid
+    magnetometerCalibration.totalBucketPoints = len(self.point_buckets)
+    if with_points:
+      magnetometerCalibration.points = self.point_buckets.get_points().tolist()
+    return msg
+
+  def compute_calibration_params(self) -> None:
+    if self.point_buckets.is_valid():
+      calibration_params = self.estimate_calibration_params()
+      if any(np.isnan(calibration_params)):
+        self.calibrationParams = CALIB_DEFAULTS
+        self.calibrated = False
+      else:
+        self.calibrationParams = calibration_params
+        self.calibrated = True
+
+
+def main(sm: Optional[messaging.SubMaster] = None, pm: Optional[messaging.PubMaster] = None):
```

Comment:
this is no longer needed and was removed recently
```suggestion
def main():
```